packages/wallet-sdk/src/core/communicator/Communicator.ts

```diff
@@ -96,5 +110,40 @@ export class Communicator {
         if (!this.popup) throw standardErrors.rpc.internal();
         return this.popup;
       });
-  }
+  };
+
+  /**
+   * Posts a RPC request to the popup window and waits for a response
+   */
+  postRPCRequest = (request: RPCRequestMessage): Promise<RPCResponseMessage> => {
+    this.pendingRequests.set(request.id, 'created');
+
+    return new Promise((resolve, reject) => {
+      // defer the execution until the current call stack clears
+      Promise.resolve().then(() => {
+        // reject if the request was cancelled between creation and awaiting due to PopupUnload
+        if (!this.pendingRequests.has(request.id)) {
+          reject(standardErrors.provider.userRejectedRequest('Request cancelled before sending'));
+          return;
+        }
+
+        // otherwise, post the message and wait for a response
+        this.pendingRequests.set(request.id, 'awaiting');
+        this.postMessage(request)
+          .then(() =>
+            this.onMessage<RPCResponseMessage>(({ requestId }) => requestId === request.id)
+          )
+          // resolve or reject the promise to take any follow-up requests if necessary before disconnecting
+          .then(resolve)
+          .catch(reject)
+          // then clean up the current request and disconnect if there are no more pending requests
+          .finally(() => {
+            this.pendingRequests.delete(request.id);
+            if (this.pendingRequests.size === 0) {
+              this.disconnect();
```

Comment:
Is it always desired behavior to have this?

**Scenario: Onboarding**
When the user is onboarding and response to the `eth_requestAccounts`, the pop-up should not close once the response is done. 

The pop-up should persist and display the onboarding success screen to link to Pano or fund.

Reply:
good point. thanks for sharing the scenarios. that helps me understand the issue better.

this line should be revisited. 
instead of calling `this.disconnect()` here (which means both sdk and popup can initiate this), `this.pendingRequests.size === 0` check should happen inside the `disconnect` method which should be triggered only by `PopupUnload` event from popup.
will take a closer look today.